Source/interactive_control.cpp

```diff
@@ -1,593 +1,615 @@
+{
+    std::shared_ptr<interactive_joystick_control> joystick = std::shared_ptr<interactive_joystick_control>(new interactive_joystick_control());
+    bool success = joystick->init_from_json(json);
+
+    if (success)
+    {
+        joystick->m_parentScene = parentSceneId;
+        interactivity_manager::get_singleton_instance()->m_impl->add_control_to_map(joystick);
+    }
+    else
+    {
+        joystick = nullptr;
+        interactivity_manager::get_singleton_instance()->m_impl->queue_interactive_event_for_client(L"Failed to initialize joystick control from json", std::make_error_code(std::errc::invalid_argument), interactive_event_type::error, nullptr);
+    }
+
+    return joystick;
+}
+
+std::shared_ptr<interactive_joystick_control>
+interactive_control_builder::build_joystick_control(
+    _In_ string_t parentSceneId,
+    _In_ string_t controlId,
+    _In_ bool enabled,
+    _In_ double x,
+    _In_ double y
+)
+{
+    std::shared_ptr<interactive_joystick_control> joystick = std::shared_ptr<interactive_joystick_control>(new interactive_joystick_control(parentSceneId, controlId, enabled, x, y));
+    interactivity_manager::get_singleton_instance()->m_impl->add_control_to_map(joystick);
+
+    return joystick;
+}
+
+//
+// Control base class
+//
+
+const interactive_control_type&
+interactive_control::control_type() const
+{
+    return m_type;
+}
+
+const string_t&
+interactive_control::control_id() const
+{
+    return m_controlId;
+}
+
+
+interactive_control::interactive_control()
+{
+
+}
+
+interactive_control::interactive_control(
+    _In_ string_t parentScene,
+    _In_ string_t controlId,
+    _In_ bool disabled
+)
+{
+    m_parentScene = std::move(parentScene);
+    m_controlId = std::move(controlId);
+    m_disabled = disabled;
+}
+
+
+bool
+interactive_button_state::is_pressed()
+{
+    return m_isPressed;
+}
+
+bool
+interactive_button_state::is_down()
+{
+    return m_isDown;
+}
+
+bool
+interactive_button_state::is_up()
+{
+    return m_isUp;
+}
+
+interactive_button_state::interactive_button_state() :
+    m_isUp(false),
+    m_isDown(false),
+    m_isPressed(false)
+{
+}
+
+uint32_t
+interactive_button_count::count_of_button_presses()
+{
+    return m_buttonPresses;
+}
+
+uint32_t
+interactive_button_count::count_of_button_downs()
+{
+    return m_buttonDowns;
+}
+
+uint32_t
+interactive_button_count::count_of_button_ups()
+{
+    return m_buttonUps;
+}
+
+void
+interactive_button_count::clear()
+{
+    m_buttonDowns = 0;
+    m_buttonPresses = 0;
+    m_buttonUps = 0;
+}
+
+interactive_button_count::interactive_button_count() :
+    m_buttonPresses(0),
+    m_buttonDowns(0),
+    m_buttonUps(0)
+{
+}
+
+double
+interactive_joystick_state::x()
+{
+    return m_x;
+}
+
+double
+interactive_joystick_state::y()
+{
+    return m_y;
+}
+
+interactive_joystick_state::interactive_joystick_state(
+    _In_ double x,
+    _In_ double y) :

```

Comment:
will fix

---
In reply to: [125970607](https://github.com/mixer/interactive-cpp/pull/17#discussion_r125970607) [](ancestors = 125970607)